Document branching strategy (#7)
inrupt/public-documentation: coding-conventions/coding-standard.md

- Appreciation for the difficult, but crucial aspects of REST; self-descriptive messages and Hypertext As The Engine Of Application State (HATEOAS).

## Branching strategy

For projects that require multiple versions of the software to be maintained in parallel,
the [Git Flow strategy](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow)

> should be followed.

Reviewer: I wasn't part of the last call, but I believe that Git Flow is vastly over-complicated. A simpler approach would be that of GitHub Flow, with the following additions:

* A \<versionNumber\>.x branch would be created to maintain a legacy release. An example of a branch name would be `14.x`.
* Master branch tracks the current development version; with the previous example, master would track `15.x`.

There is no `develop` branch in this case; `master` is the development branch, and git tags would be used to tag releases, as we'd use in GitHub Flow. This keeps our two git strategies closely aligned, with the difference that software with multiple major versions in active development may have multiple long-lived branches. Less context switching per codebase keeps things simple.

Author: For the new open-source Solid server, GitHub Flow would indeed be perfect, I agree that Git Flow would be overkill for that. My reasoning for including it in this document was that some projects may require maintenance of multiple versions in parallel, for which GitHub Flow would be better. No strong opinion on this though, so happy to take it out and go all-in on (modified) GitHub Flow.

> A <versionNumber>.x branch would be created to maintain a legacy release. An example of a branch name would be 14.x.

Would you make this required or optional? For projects that only maintain a single version, this may introduce many unnecessary branches, for which tags may be better suited.

Reviewer: > Would you make this required or optional? For projects that only maintain a single version, this may introduce many unnecessary branches, for which tags may be better suited.

This would specifically be used in projects that require multiple maintained versions; master always tracks the latest stable, released major version.

The same strategy should work for major refactors in which a future major release may need to be created; if the current stable version is 15.x, and a change is planned which would require a major version change, a `16.x` branch would be created (and then deleted once merged with master, which would then track 16.x.)

Otherwise, the simpler [GitHub Flow strategy](https://guides.github.com/introduction/flow/) can be followed.

**Git Flow:**

- Master branch is a representation of the software releases.
- Development branch is where all features and bugfixes are merged into. This branch is merged into master upon each release.
- Feature branches are created for every separate feature (including fixes) that is being developed, which are merged into the development branch once the feature is completed (via a PR).

**GitHub Flow:**

> - Master branch is the main development branch where all features and bugfixes are merged into. At any point in time, it should be deployable.

Reviewer: GitHub flow doesn't specify anything about history, merging, and rebasing, but it may be worth adding that not only should master always be deployable, but _any_ commit in master should be deployable; that is, each commit should contain an atomic, revertable which itself only fixes things or adds features, rather than partially-broken code.

In normal development (or at least mine), the natural history may look like:

* `Refactor button to allow custom styles`
* `Add fancy button component`
* `bugfix: allow green buttons again`
* `Restructure button files and add tests`

Where, upon final approval, it may be rebased:

* `Refactor button to allow custom styles, restructure files, and add tests`
* `Add fancy button component`

This would allow the fancy button to be reverted while still maintaining the refactor. A doesn't always have to squash its commits, but the commits should be as few as possible while remaining descriptive.

Author: 👍 

I expect that this would align well with ConventionalCommits, and enable straightforward changelog generation.

- Feature branches are created for every separate feature (including fixes) that is being developed, which are merged into the master branch once the feature is completed (via a PR).
- In order to link the state of the code to software releases, the GitHub releases/tags feature can be used.

> In both cases, feature branches must have a descriptive name, such as `fix/broken-homepage` or `feature/fancy-button`.

Reviewer 2: Doesn't ConventionalCommit's REQUIRE `feat/` and not `feature/`...?

Author: AFAIK ConventionalCommits only defines commit naming, but not branch naming.

Nevertheless, I think it's a good idea to align the branch naming strategy to ConventionalCommits' commit types, so I'm happy to add this in this PR once a formal decision has been made.

Reviewer: I'm a big fan of ConventionalCommit. It'd also be nice to solidify branch naming, although I don't have strong opinions on the format, and I don't think it's nearly as important as the commit message format, as long as the branch name is descriptive. `fix/`, and `feature/` or `feat/` seem reasonable.

- Appreciation for the difficult, but crucial aspects of REST; self-descriptive messages and Hypertext As The Engine Of Application State (HATEOAS).

## Branching strategy

By default, the [GitHub Flow strategy](https://guides.github.com/introduction/flow/) should be followed:

- Master branch is the main development branch where all features and bugfixes are merged into. At any point in time, it should be deployable.
- Feature branches are created for every separate feature (including fixes) that is being developed, which are merged into the master branch once the feature is completed (via a PR).

> - In order to link the state of the code to software releases, the GitHub releases/tags feature can be used.

Reviewer 3: I'd prefer (to at least allow) the use of Git tags, because that will make the state of the code of releases (and their release notes) available independently of GitHub.

```suggestion
- In order to link the state of the code to software releases, the GitHub releases/Git tags feature can be used.
```

Reviewer: Should tags be emphasized in that case?

```suggestion
+ In order to link the state of the code to software releases, Git tags should be used to track releases, and GitHub releases can be used to help document releases. Tags should be in the format `vX.X.X`, such as `v1.13.1`.
```

Reviewer 3: Sounds good, though I'd prefer to use a `CHANGELOG` file to document releases for the same reason. (Though of course, it can be copied over to GitHub releases as a courtesy.)

Reviewer: Right - GitHub releases are there to help, not to replace the existence of tags and a `CHANGELOG`. Do you think we should explain that here as well?

Reviewer 3: Yeah, it reads as if they're the primary source of truth now, to me. How about this:

```suggestion
- In order to link the state of the code to software releases, Git tags should be used to track releases, and the changelog to help document releases. Optionally, the changelog update can also be recorded in the GitHub Release. Tags should be in the format `vX.X.X`, such as `v1.13.1`.
```

Though the whole optional part can also be removed as far as I'm concerned, given that that's pretty much implied. Unless someone knows of a good reason to mandate a GitHub Release page.

Reviewer: Using the release page makes sense when people are downloading built artifacts, but not necessarily for libraries published to package managers. I'd be okay with leaving it off.

Reviewer: ```suggestion
- In order to link the state of the code to software releases, Git tags should be used to track releases, and a CHANGELOG file should document releases. Tags should be in the format `vX.X.X`, such as `v1.13.1`.
```

For projects that require the maintenance of multiple versions in parallel,
the following additions to this strategy can be made:

> - Long-lived version branches are created to maintain legacy release. For example, an old version 14.x can be maintained in branch `version/14.x`.

Reviewer 3: Given that every release should be tagged, it seems to me to be unnecessary overhead to also create branches for each of those releases. Additionally, we should make sure that these fixes also get included in later releases.

```suggestion
- Version-specific branches can be created when preparing an update to a legacy release. For example, if version 15 is the most recent release, a patch to version 14.x can be prepared in branch `version/14.x`. After release, this branch should be merged back into `master`.
```

**Note: Version branches should be created for major refactors in a future version, which should be merged into master as soon as possible. For example, if the current stable version is 15.x, and a change is planned which would require a major version change, a 16.x branch would be created, and then deleted once merged with master, which would then track 16.x.**

In both cases, feature branches must have a descriptive name following the format `type/short-description`,
where types from [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) can be used.
Examples of branch names are: `fix/broken-homepage`, `feat/fancy-button`.

Upon merging any branch into master, it should be ensured that not only the master branch is deployable, but *every commit* is deployable as well.
This means that every commit should be atomic, which means that each commit adds or fixes a specific thing, instead of containing partially-broken code.
This also ensures that each commit is revertible at any time.

> Non-master branches can contain non-atomic commits for work in progress, but must then be rebased as soon as possible.

Reviewer 3: Rebasing as quickly as possible can cause extra for people who have checked out the code while reviewing, so best to align that with the reviewers.

```suggestion
Non-master branches can contain non-atomic commits for work in progress, but must then be rebased before merging to master.
```
